src/Rewards.sol

```diff
@@ -0,0 +1,181 @@
+ * This contract must have the MINTER_ROLE in the NODL token contract.
+ */
+contract Rewards is AccessControl, EIP712 {
+    /**
+     * @dev Role required to set the reward quota.
+     */
+    bytes32 public constant QUOTA_SETTER_ROLE = keccak256("QUOTA_SETTER_ROLE");
+
+    /**
+     * @dev The signing domain used for generating signatures.
+     */
+    string public constant SIGNING_DOMAIN = "rewards.depin.nodle";
+
+    /**
+     * @dev The version of the signature scheme used.
+     */
+    string public constant SIGNATURE_VERSION = "1";
+
+    /**
+     * @dev This constant defines the reward type.
+     * This should be kept consistent with the Reward struct.
+     */
+    bytes public constant REWARD_TYPE = "Reward(address recipient,uint256 amount,uint256 counter)";
+
+    /**
+     * @dev Reference to the NODL token contract.
+     */
+    NODL public nodlToken;
+
+    /**
+     * @dev Maximum amount of rewards that can be distributed in a period.
+     */
+    uint256 public rewardQuota;
+
+    /**
+     * @dev Duration of each reward period.
+     */
+    uint256 public rewardPeriod;
+
+    /**
+     * @dev Timestamp indicating when the reward quota is due to be renewed.
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward counters for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardCounters;
+
+    /**
+     * @dev Struct on which basis an individual reward must be issued.
+     */
+    struct Reward {
+        address recipient;
+        uint256 amount;
+        uint256 counter;
+    }
+
+    /**
+     * @dev Error when the reward quota is exceeded.
+     */
+    error RewardQuotaExceeded();
+
+    /**
+     * @dev Error when the reward is not from the authorized oracle.
+     */
+    error UnauthorizedOracle();
+
+    /**
+     * @dev Error when the recipient's counter does not match.
+     */
+    error InvalidRecipientCounter();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param nodlTokenAddress Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(address nodlTokenAddress, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
```

Comment:
```suggestion
    constructor(NODL nodlTokenAddress, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
```
avoids the type conversion below :)

Reply:
That was my preference to make it clear it's the address given in the constructor, but I am flexible to do the other way.

Reply:
The other benefit of passing an address is when we decouple Rewards from NODL contract use and interface for an ERC20 mintable instead. This might not be needed now but points us to which method should be our habitual method.

Reply:
> The other benefit of passing an address is when we decouple Rewards from NODL contract and instead use an interface for an ERC20 mintable instead. This might not be needed now but points us to which method should be our habitual method.

Then we will use an interface contract when the time comes. That would be the more standard way to do it